File: app/src/main/java/io/github/ulysseszh/displaytoggle/activity/ScreenController.java
package com.example.screenoff;

import android.os.IBinder;
import java.lang.reflect.Method;

public class ScreenController {

    public static void turnOff() {
        try {
            Class<?> sc = Class.forName("android.view.SurfaceControl");

            long[] ids = (long[]) sc.getMethod("getPhysicalDisplayIds").invoke(null);
            if (ids == null) return;

            Method getToken = sc.getMethod("getPhysicalDisplayToken", long.class);
            Method setPower = sc.getMethod("setDisplayPowerMode", IBinder.class, int.class);

            for (long id : ids) {
                IBinder token = (IBinder) getToken.invoke(null, id);
                setPower.invoke(null, token, 0);
            }
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    public static void turnOn() {
        try {
            Class<?> sc = Class.forName("android.view.SurfaceControl");

            long[] ids = (long[]) sc.getMethod("getPhysicalDisplayIds").invoke(null);
            if (ids == null) return;

            Method getToken = sc.getMethod("getPhysicalDisplayToken", long.class);
            Method setPower = sc.getMethod("setDisplayPowerMode", IBinder.class, int.class);

            for (long id : ids) {
                IBinder token = (IBinder) getToken.invoke(null, id);
                setPower.invoke(null, token, 2);
            }
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/io/github/ulysseszh/displaytoggle/activity/ScreenController.kt
package io.github.ulysseszh.displaytoggle.activity // Thống nhất dùng package này

import android.os.IBinder
import java.lang.reflect.Method

object ScreenController {
    // THÊM 2 DÒNG NÀY VÀO
    const val OFF = 0
    const val ON = 2

    fun turnOff() {
        setPowerMode(OFF)
    }

    fun turnOn() {
        setPowerMode(ON)
    }

    // Viết gọn lại hàm setPower để tái sử dụng hằng số
    private fun setPowerMode(mode: Int) {
        try {
            val sc = Class.forName("android.view.SurfaceControl")
            val ids = sc.getMethod("getPhysicalDisplayIds").invoke(null) as? LongArray ?: return
            val getToken: Method = sc.getMethod("getPhysicalDisplayToken", Long::class.javaPrimitiveType)
            val setPower: Method = sc.getMethod("setDisplayPowerMode", IBinder::class.java, Int::class.javaPrimitiveType)

            for (id in ids) {
                val token = getToken.invoke(null, id) as IBinder
                setPower.invoke(null, token, mode)
            }
        } catch (e: Throwable) {
            e.printStackTrace()
        }
    }
}
